test/pylib/scylla_cluster.py

```diff
@@ -382,20 +391,30 @@ async def start(self, api: ScyllaRESTAPIClient) -> None:
                         logpath = log_handler.baseFilename   # type: ignore
                     else:
                         logpath = "?"
-                    raise RuntimeError(f"Failed to start server at host {self.ip_addr}.\n"
+                    raise RuntimeError(f"Failed to start server with ID = {self.server_id}, IP = {self.ip_addr}.\n"
                                        "Check the log files:\n"
                                        f"{logpath}\n"
                                        f"{self.log_filename}")
 
             if hasattr(self, "host_id") or await self.get_host_id(api):
-                if await self.cql_is_up():
+                cql_up_state = await self.cql_is_up()
+                if cql_up_state == CqlUpState.QUERIED:
                     return
 
             # Sleep and retry
             await asyncio.sleep(sleep_interval)
 
-        raise RuntimeError(f"failed to start server {self.ip_addr}, "
-                           f"check server log at {self.log_filename}")
+        err = f"Failed to start server with ID = {self.server_id}, IP = {self.ip_addr}."
+        if hasattr(self, "host_id"):
+            err += f" Managed to obtain the server's Host ID ({self.host_id})"
+            if cql_up_state == CqlUpState.CONNECTED:
+                err += " and to connect the CQL driver, but failed to execute a query."
+            else:
+                err += " but failed to connect the CQL driver."
+        else:
+            err += " Failed to obtain the server's Host ID."
+        err += f"\nCheck server log at {self.log_filename}."
+        raise RuntimeError(err)
```

Comment:
I think your approach is valid, but I think you could have done something simpler: cql_is_up could have thrown an exception (say NotYetUp) when Scylla is not yet up, the exception saying what went wrong. Then in a loop you call cql_is_up retrying on NotYetUp - and finally on timeout call cql_is_up() again just to get it to throw the exception for the user to see it.

I used that approach in test/cql-pytest/run.py.

Reply:
I don't much like returning values through exceptions; `cql_is_up` answers the question "is CQL up?" - it would be awkward to return "yes" through a value and "no" through an exception, it's not an exceptional situation, it's expected.

Reply:
Ok, I guess it's a matter of style. I think the current design makes your code longer and also more spread out (one place checks if we're done, a different place reports why exactly we're not done using various flags that the first place left for us). But it's completely a matter of taste, so I am not insisting at all.